query-service-impl/src/main/java/org/hypertrace/core/query/service/prometheus/PrometheusBasedRequestHandler.java

```diff
@@ -74,13 +91,74 @@ public QueryCost canHandle(QueryRequest request, ExecutionContext executionConte
   @Override
   public Observable<Row> handleRequest(
       QueryRequest originalRequest, ExecutionContext executionContext) {
-
     // Validate QueryContext and tenant id presence
     Preconditions.checkNotNull(executionContext);
     Preconditions.checkNotNull(executionContext.getTenantId());
 
-    // todo call convert and execute request using client here
+    Map<Request, PromQLMetricResponse> responseMap;
+    Map<String, String> metricNameToQueryMap = new LinkedHashMap<>();
+    if (isRangeQueryRequest(originalRequest)) {
+      PromQLRangeQueries promQLRangeQueries =
+          requestToPromqlConverter.convertToPromqlRangeQuery(
+              executionContext,
+              originalRequest,
+              executionContext.getAllSelections(),
+              metricNameToQueryMap);
+      responseMap = prometheusRestClient.executeRangeQuery(promQLRangeQueries);
+    } else {
+      PromQLInstantQueries promQLInstantQueries =
+          requestToPromqlConverter.convertToPromqlInstantQuery(
+              executionContext,
+              originalRequest,
+              executionContext.getAllSelections(),
+              metricNameToQueryMap);
+      responseMap = prometheusRestClient.executeInstantQuery(promQLInstantQueries);
+    }
+
+    List<Row> rows =
+        PrometheusBasedResponseBuilder.buildResponse(
+            responseMap,
+            prometheusViewDefinition.getAttributeMap(),
+            metricNameToQueryMap,
+            prepareColumnSet(executionContext.getAllSelections(), executionContext),
+            executionContext.getTimeFilterColumn());
+
+    return Observable.fromIterable(rows).doOnNext(row -> LOG.debug("collect a row: {}", row));
+  }
+
+  private boolean isRangeQueryRequest(QueryRequest queryRequest) {
+    return queryRequest.getGroupByList().stream().anyMatch(QueryRequestUtil::isDateTimeFunction);
+  }
 
-    return null;
+  private LinkedHashSet<String> prepareColumnSet(
+      LinkedHashSet<Expression> expressions, ExecutionContext executionContext) {
+    LinkedHashSet<String> columnSet = new LinkedHashSet<>();
+
+    expressions.forEach(
+        expression -> {
+          ValueCase valueCase = expression.getValueCase();
+          switch (valueCase) {
+            case ATTRIBUTE_EXPRESSION:
+            case COLUMNIDENTIFIER:
+              columnSet.add(
+                  QueryRequestUtil.getLogicalColumnNameForSimpleColumnExpression(expression));
+              break;
+            case FUNCTION:
+              if (QueryRequestUtil.isDateTimeFunction(expression)) {
+                columnSet.add(executionContext.getTimeFilterColumn());
+              } else {
+                Function function = expression.getFunction();
+                String columnName =
+                    QueryRequestUtil.getLogicalColumnNameForSimpleColumnExpression(
+                        function.getArguments(0));
```

Comment:
Can we have a common function for forming logical name `<function_name>:<attributeId>`?